### PR TITLE
Reimagine Sharing: Add scrollable timeline view

### PR DESCRIPTION
**Pull request by bjtitus**
| 📘 Part of: https://github.com/Automattic/pocket-casts-ios/issues/1910 |
|:---:|

Adds the timeline view for clip sharing from the Reimagine Sharing project.

https://github.com/Automattic/pocket-casts-ios/assets/3250/d7902514-2999-427b-b0b3-17d48e3b540e

### Code Changes

* `MediaTrimBar` contains the play button and the timeline with associated styled elements
* `MediaTrimView` contains the scrollable timeline portion which will include Trim handles in the future
* `ScrollableScrollView` is a wrapper which handles scrolling operations via the `scrollTo` modifier and set of `invisibleTickMarks`. There are some newer APIs we could use in the future but this approach works for now.
* `AudioWaveformView` renders the lines for our "pseudo waveform". This may be replaced with an actual waveform in the future but is a static stylized version for now.
* `PCUnevenRoundedRectangle` is a fallback for rounding corners in iOS 15 similar to `UnevenRoundedRectangle` in iOS 16+.

## To test

> [!NOTE]
> Enable the `newSharing` feature flag to test this

* Open the Now Playing view in the player
* Tap the share button in the upper right
* Tap the "Clip" option
* Verify that the Clip sharing page appears
* Ensure that timeline is scrolled to the correct position
* Verify that the timeline behaves as expected by scrolling back and forth and changing playback position from the episode and re-entering

A few missing items which I'll submit separately:
* Zoom gesture handling - the building blocks are there in `AudioWaveformView ` for hiding and showing lines
* Playback following
* Play/Pause button functionality

The tick marks may not match the designs exactly as they will be recalculated depending on scale (which may adjust with zooming). I'll be cleaning this up in a future update.

## Checklist

- [x] I have considered if this change warrants user-facing release notes and have added them to `CHANGELOG.md` if necessary.
- [x] I have considered adding unit tests for my changes.
- [x] I have updated (or requested that someone edit) [the spreadsheet](https://docs.google.com/spreadsheets/d/107jqrutZhU0fVZJ19SBqxxVKbV2NWSdQC9MFYdLiAxc/edit?usp=sharing) to reflect any new or changed analytics.

### CODE REVIEW

File: Modules/Utils/Sources/PocketCastsUtils/Extensions/ComparableExtension.swift [hunk 1]
```diff
@@ -9,3 +9,13 @@ public extension Comparable {
         self > limits.lowerBound && self < limits.upperBound ? self : self.clamped(to: limits)
     }
 }
+
+public extension Comparable {
+    func clamped(to limits: ClosedRange<Self>) -> Self {
```
Comment (bjtitus) [inline Modules/Utils/Sources/PocketCastsUtils/Extensions/ComparableExtension.swift:14]: These are duplicates of the `Range` versions for `ClosedRange`. Unfortunately, they don't share a suitable protocol so I couldn't make these generic.

File: podcasts/Common SwiftUI/RoundedCorner.swift [hunk 2]
```diff
@@ -0,0 +1,70 @@
+import SwiftUI
+
+/// Used as a fallback for iOS 15 when `UnevenRoundedRectangle` is necessary
```
Comment (leandroalonso) [inline podcasts/Common SwiftUI/RoundedCorner.swift:3]: 🥲

File: podcasts/Sharing/AudioWaveformView.swift [hunk 3]
```diff
@@ -0,0 +1,112 @@
+import SwiftUI
+
+struct AudioWaveformView: View {
+    let baseLineWidth: CGFloat = 2
+    let baseLineSpacing: CGFloat = 2
+    var scale: CGFloat
+    var width: CGFloat
```
Comment (leandroalonso) [inline podcasts/Sharing/AudioWaveformView.swift]: [nit] Do these need to be exposed?

File: podcasts/Sharing/Clip/PlayButton.swift [hunk 4]
```diff
@@ -0,0 +1,19 @@
+import SwiftUI
+
+struct PlayButton: View {
```
Comment (leandroalonso) [inline podcasts/Sharing/Clip/PlayButton.swift]: Maybe rename it to `TrimPlayButton`?

I can see confusion when searching for the current `PlayPauseButton` in the full player.

**Review by leandroalonso — APPROVED** [review]
Tested on real deice and on multiple devices. On iPad there's a small issue, the bottom padding is missing:

Simulator Screenshot - iPad Pro (12 9-inch) (6th generation) - 2024-07-11 at 11 07 43

It's not a blocker though and can be addressed in the future.

**Comment by bjtitus** [comment]
@leandroalonso Good catch! This was an issue on all of the new sharing views. I added a change (https://github.com/Automattic/pocket-casts-ios/pull/1896/commits/c6f89188bc501474b1c158700f928b1d22badc75) to pad both top and bottom on iPad.

**Pull request merged**